irohad/ametsuchi/impl/postgres_query_executor.cpp

```diff
@@ -0,0 +1,800 @@
+      auto result = boost::apply_visitor(visitor_, query.get());
+      return clone(result.queryHash(query.hash()).build());
+    }
+
+    bool PostgresQueryExecutor::validate(
+        const shared_model::interface::BlocksQuery &query) {
+      boost::format cmd(R"(%s)");
+      soci::statement st =
+          (sql_->prepare
+           << (cmd % checkAccountRolePermission(Role::kGetBlocks)).str());
+      int has_perm;
+      st.exchange(soci::use(query.creatorAccountId(), "role_account_id"));
+      st.exchange(soci::into(has_perm));
+      st.define_and_bind();
+      st.execute(true);
+
+      return has_perm > 0;
+    }
+
+    PostgresQueryExecutorVisitor::PostgresQueryExecutorVisitor(
+        soci::session &sql,
+        std::shared_ptr<shared_model::interface::CommonObjectsFactory> factory,
+        KeyValueStorage &block_store)
+        : sql_(sql), block_store_(block_store), factory_(factory) {}
+
+    void PostgresQueryExecutorVisitor::setCreatorId(
+        const shared_model::interface::types::AccountIdType &creator_id) {
+      creator_id_ = creator_id;
+    }
+
+    QueryResponseBuilderDone PostgresQueryExecutorVisitor::operator()(
+        const shared_model::interface::GetAccount &q) {
+      using T = boost::tuple<
+          boost::optional<shared_model::interface::types::AccountIdType>,
+          boost::optional<shared_model::interface::types::DomainIdType>,
+          boost::optional<shared_model::interface::types::QuorumType>,
+          boost::optional<shared_model::interface::types::DetailType>,
+          boost::optional<shared_model::interface::types::RoleIdType>,
+          int>;
+      T tuple;
+      auto cmd = (boost::format(R"(WITH has_perms AS (%s),
+      t AS (
+          SELECT a.account_id, a.domain_id, a.quorum, a.data, ARRAY_AGG(ar.role_id) AS roles
+          FROM account AS a, account_has_roles AS ar
+          WHERE a.account_id = :target_account_id
+          AND ar.account_id = a.account_id
+          GROUP BY a.account_id
+      )
+      SELECT account_id, domain_id, quorum, data, roles, perm
+      FROM t RIGHT OUTER JOIN has_perms AS p ON TRUE
+      )")
+                  % hasQueryPermission(creator_id_,
+                                       q.accountId(),
+                                       Role::kGetMyAccount,
+                                       Role::kGetAllAccounts,
+                                       Role::kGetDomainAccounts))
+                     .str();
+      soci::statement st = (sql_.prepare << cmd);
+      st.exchange(soci::use(q.accountId(), "target_account_id"));
+      st.exchange(soci::into(tuple));
+
+      st.define_and_bind();
+      st.execute(true);
+
+      if (not tuple.get<5>()) {
+        return statefulFailed();
+      }
+
+      if (not tuple.get<0>().is_initialized()) {
+        return buildError<shared_model::interface::NoAccountErrorResponse>();
+      }
+
+      auto roles_str = tuple.get<4>().get();
```

Comment:
Please add check for optional

Reply:
It is not required here, because either all optional fields are set, or there is no permission.